hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1148,31 +1148,39 @@ def example_deletion_with_block_lowering(self):
         example and every block not inside that example it tries deleting the
         example and modifying the block's value by one in either direction.
         """
-        i = 0
-        while i < len(self.shrink_target.blocks):
-            if not self.is_shrinking_block(i):
-                i += 1
-                continue
+        example_index = 0
 
-            u, v = self.blocks[i].bounds
-
-            j = 0
-            while j < len(self.shrink_target.examples):
+        while True:
+            any_examples_checked = False
+            i = 0
+            while i < len(self.shrink_target.blocks):
+                u, v = self.blocks[i].bounds
                 n = int_from_bytes(self.shrink_target.buffer[u:v])
-                if n == 0:
-                    break
-                ex = self.shrink_target.examples[j]
-                if ex.start < v or ex.length == 0:
-                    j += 1
-                    continue
 
-                buf = bytearray(self.shrink_target.buffer)
-                buf[u:v] = int_to_bytes(n - 1, v - u)
-                del buf[ex.start : ex.end]
-                if not self.incorporate_new_buffer(buf):
-                    j += 1
+                if n == 0 or not self.is_shrinking_block(i):
+                    i += 1
+                    continue
 
-            i += 1
+                examples = [
+                    ex
+                    for ex in self.shrink_target.examples
+                    if ex.start >= v and ex.length > 0
+                ]
+                examples.sort(key=lambda ex: ex.depth)
```

Comment:
Filtering and sorting the example list on every inner iteration seems bold, though I don't have a good intuition for whether it's actually a problem in practice.

If we assume that most iterations will fail to shrink, then it could be a good idea to have a local helper function that stores a memoized copy of the full depth-sorted example list, and then automatically invalidates and rebuilds the list when it detects that a shrink has occurred since last time.

Reply:
Yeah, fair point. I think it's probably *not* a problem in practice given the relative costs - there are only `O(n log(n))` examples and building a new test costs `O(n)` anyway, but it's probably strictly better to not do it.